### PR TITLE
Implement SYCL TeamPolicy for vector_size > 1

### PR DESCRIPTION
**Pull request by masterleinad**
Most of the implementation is mirrored from `CUDA`/`HIP`. For `SYCL` kernels, we don't know the subgroup (warp size) at runtime; we can only obtain a vector of possible values from which we set the maximal vector size. In the end, the requirement is that the subgroup size is divisible by the vector size and this pull request adds a check when compiling with `Kokkos_ENABLE_DEBUG`.
Also, some of the tests needed to be adapted for Intel hardware since the maximum workgroup size is limited by 256.

### CODE REVIEW

**Comment by masterleinad** [comment]
This seems to also work on NVIDIA GPUs now. I don't have a good explanation for the deadlock in the subgroup barrier in `Test12a_ThreadScratch` and `Test12b_TeamScratch`, though. The corresponding vector range is executed with `vector_size==1` anyway so replacing it with a serial loop should not hurt too much as a workaround.

**Comment by masterleinad** [comment]
Also, we can't specify a width for the sub-group shuffle operations, which means that we have to do some more index calculation ourselves compared to the corresponding `CUDA` instructions.

File: core/unit_test/TestTeam.hpp [hunk 1]
```diff
@@ -970,7 +970,11 @@ struct ClassNoShmemSizeFunction {
                 double *, ExecSpace,
                 Kokkos::MemoryTraits<Kokkos::Unmanaged> >::shmem_size(1600);
 
-    int team_size = 8;
+#ifdef KOKKOS_ENABLE_SYCL
+    int team_size = 4;
+#else
+    int team_size      = 8;
```
Comment (dalg24) [inline core/unit_test/TestTeam.hpp:976]: The formatting looks interesting.

File: core/unit_test/incremental/Test12a_ThreadScratch.hpp [hunk 2]
```diff
@@ -74,9 +74,15 @@ struct ThreadScratch {
     for (int i = 0; i < sY; ++i) v_S(i) = 0;
 
     Kokkos::parallel_for(Kokkos::TeamThreadRange(team, sX), [&](const int m) {
+    // FIXME_SYCL This deadlocks in the subgroup_barrier when running on CUDA
+    // devices.
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/unit_test/incremental/Test12a_ThreadScratch.hpp:78]: I would expect it to be an issue when compiling for the `CUDA` backend in `SYCL` (and thus the generated code) rather than that it's related to `NVIDIA` GPU's directly.
Either way, I am happy to change the wording if that is preferred.

File: core/src/SYCL/Kokkos_SYCL.cpp [hunk 3]
```diff
@@ -224,10 +224,6 @@ std::ostream& SYCL::impl_sycl_info(std::ostream& os,
             << device.get_info<device::global_mem_cache_size>()
             << "\nGlobal Mem Size: "
             << device.get_info<device::global_mem_size>()
-            << "\nMax Constant Buffer Size: "
-            << device.get_info<device::max_constant_buffer_size>()
-            << "\nMax Constant Args: "
-            << device.get_info<device::max_constant_args>()
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL.cpp:230]: Yes, this pull request originally contained all the changes to avoid using deprecated code wrt to the compiler version I was using to compose this pull request in (separate commits, see [`00b5531` (#4183)](https://github.com/kokkos/kokkos/pull/4183/commits/00b553153adf220ca8251e68b99b10cb9e66db06), [`234d6f2` (#4183)](https://github.com/kokkos/kokkos/pull/4183/commits/234d6f2f7912e8430efe04f44e3bde7d41d1d843), https://github.com/kokkos/kokkos/pull/4183/commits/2557434904629fd935b17b60a4a6fe222578f400). Most of them I needed to revert again, see [`2de95e2` (#4183)](https://github.com/kokkos/kokkos/pull/4183/commits/2de95e29ada9bdc2bf9647f4bd2183e7650a343f), [`fea18fd` (#4183)](https://github.com/kokkos/kokkos/pull/4183/commits/fea18fd530bc40335be64ac4bb5268030356c72c), since the version we are testing with doesn't support the new names yet.

I am currently preferring to perform these updates as drive-by so that I don't need to keep track of them all the time when working on multiple pull requests. Anyway, I can remove them if desired.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp [hunk 4]
```diff
@@ -129,24 +127,23 @@ class TeamPolicyInternal<Kokkos::Experimental::SYCL, Properties...>
   }
   inline bool impl_auto_vector_length() const { return m_tune_vector_length; }
   inline bool impl_auto_team_size() const { return m_tune_team_size; }
-  static int vector_length_max() {
-    // FIXME_SYCL provide a reasonable value
-    return 1;
+  int vector_length_max() const {
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp:130]: Hmm... good point.

File: core/src/SYCL/Kokkos_SYCL_Team.hpp [hunk 5]
```diff
@@ -287,19 +285,43 @@ class SYCLTeamMember {
...
+
+    const auto sg = m_item.get_sub_group();
+
+    if (grange1 == 1) return;
+
+    // Intra vector lane shuffle reduction:
+    typename ReducerType::value_type tmp(value);
+    typename ReducerType::value_type tmp2 = tmp;
+
+    for (int i = grange1; (i >>= 1);) {
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Team.hpp:310]: Yes, I stole the implementation from there.

File: core/unit_test/TestTeam.hpp (hunk 1 above)
Comment (masterleinad) [reply core/unit_test/TestTeam.hpp:976]: I know but I couldn't quite figure out why (and it didn't seem to be too important to me).

File: core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp:130]: https://github.com/kokkos/kokkos/pull/4186 is an attempt to work around this issue.

File: core/src/SYCL/Kokkos_SYCL.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (dalg24) [reply core/src/SYCL/Kokkos_SYCL.cpp:230]: Now that's fine by me

File: core/unit_test/incremental/Test12a_ThreadScratch.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (dalg24) [reply core/unit_test/incremental/Test12a_ThreadScratch.hpp:78]: So that is SYCL terminology and not Kokkos, is that right?

[the remark this answers is not in the document]
Comment (masterleinad) [reply core/unit_test/incremental/Test12a_ThreadScratch.hpp:78]: I see more `CUDA` than `NVIDIA` in git@github.com/intel/llvm if that is the question. The specifications don't mention any of that, of course.
In this context, I use `CUDA` and `NVIDIA` GPUs as synonyms. Again, I'm happy to change this WIP comment.

**Review by Rombur — APPROVED** [review]
Looks good to me given that we fix the `static` problem  for `vector_length_max`.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp]: I went with using the default execution space for querying the maximum vector length now to disentangle this pull request from #4168. This should give the correct value except for corner cases.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp [hunk 6]
```diff
@@ -129,30 +127,34 @@ class TeamPolicyInternal<Kokkos::Experimental::SYCL, Properties...>
...
+        execution_space{}
+            .impl_internal_space_instance()
+            ->m_queue->get_device()
+            .template get_info<sycl::info::device::sub_group_sizes>();
+    return *std::max_element(sub_group_sizes.begin(), sub_group_sizes.end());
   }
 
-  static int verify_requested_vector_length(int requested_vector_length) {
+ private:
+  int verify_requested_vector_length(int requested_vector_length) const {
```
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp:142]: can be static with your recent changes

File: core/unit_test/incremental/Test12b_TeamScratch.hpp [hunk 7]
```diff
@@ -68,7 +68,7 @@ struct TeamScratch {
 
     Kokkos::parallel_for(
         "Team",
-        policy_t(pN, Kokkos::AUTO)
+        policy_t(pN, Kokkos::AUTO, 1)
```
Comment (dalg24) [inline core/unit_test/incremental/Test12b_TeamScratch.hpp:71]: Did you mean to change the vector length?

File: core/unit_test/incremental/Test12a_ThreadScratch.hpp [hunk 8]
```diff
@@ -93,7 +99,7 @@ struct ThreadScratch {
     int scratchSize = scratch_t::shmem_size(sY);
     // So this works with deprecated code enabled:
     policy_t policy =
-        policy_t(pN, Kokkos::AUTO)
+        policy_t(pN, Kokkos::AUTO, 1)
```
Comment (dalg24) [inline core/unit_test/incremental/Test12a_ThreadScratch.hpp:102]: Did you mean that change?  Why is this ok?

File: core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp (hunk 6 above)
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp:142]: Yes, it can be at the moment. Do you prefer I change this again?

File: core/unit_test/incremental/Test12a_ThreadScratch.hpp (hunk 8 above)
Comment (masterleinad) [reply core/unit_test/incremental/Test12a_ThreadScratch.hpp:102]: It doesn't change anything. The default vector length is 1 (see https://github.com/kokkos/kokkos/blob/9cf0d5bc02513aece0a79c12d717714cf9345700/core/src/Cuda/Kokkos_Cuda_Parallel.hpp#L288-L291, https://github.com/kokkos/kokkos/blob/9cf0d5bc02513aece0a79c12d717714cf9345700/core/src/HIP/Kokkos_HIP_Parallel_Team.hpp#L239-L242, and https://github.com/kokkos/kokkos/blob/9cf0d5bc02513aece0a79c12d717714cf9345700/core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp#L203-L206) and I wanted to stress that to justify replacing `ThreadVectorRange` with a serial loop.

File: core/unit_test/incremental/Test12b_TeamScratch.hpp (hunk 7 above)
Comment (masterleinad) [reply core/unit_test/incremental/Test12b_TeamScratch.hpp:71]: See above. 1 is the default.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp (hunk 6 above)
Comment (dalg24) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp:142]: Yes

Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Team.hpp]: Fixed.

File: core/src/SYCL/Kokkos_SYCL_Team.hpp [hunk 9]
```diff
@@ -287,19 +285,43 @@ class SYCLTeamMember {
...
+      tmp2 = sg.shuffle_down(tmp, i);
```
Comment (crtrott) [inline core/src/SYCL/Kokkos_SYCL_Team.hpp:311]: is this guaranteed to be a subgroup barrier?

File: core/src/SYCL/Kokkos_SYCL_Team.hpp [hunk 10]
```diff
@@ -785,21 +870,26 @@ template <class FunctorType>
...
 template <class FunctorType, class ValueType>
 KOKKOS_INLINE_FUNCTION void single(
     const Impl::VectorSingleStruct<Impl::SYCLTeamMember>& single_struct,
     const FunctorType& lambda, ValueType& val) {
-  if (single_struct.team_member.item().get_local_id(1) == 0) lambda(val);
+  const sycl::nd_item<2> item = single_struct.team_member.item();
+  const auto grange1          = item.get_local_range(1);
+  const auto sg               = item.get_sub_group();
+  if (item.get_local_id(1) == 0) lambda(val);
+  val = sg.shuffle(val, (sg.get_local_id() / grange1) * grange1);
```
Comment (crtrott) [inline core/src/SYCL/Kokkos_SYCL_Team.hpp:884]: i don't get this. Shouldn't item.get_local_id(1)==sg.get_local_id()? and shouldn't 0<=sg.get_local_id()<grange1 be true too?

File: core/unit_test/incremental/Test12a_ThreadScratch.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (crtrott) [reply core/unit_test/incremental/Test12a_ThreadScratch.hpp:78]: You should open an issue on this since a deadlock indicated a bug in the code or in the instructions generated.

**Review by crtrott — CHANGES_REQUESTED** [review]
Got a couple questions.

File: core/src/SYCL/Kokkos_SYCL_Team.hpp (hunk 9 above)
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Team.hpp:311]: https://www.khronos.org/registry/OpenCL/extensions/intel/cl_intel_subgroups.html says:

> The Intel extension adds a rich set of subgroup "shuffle" functions to allow work items within a work group to interchange data without the use of local memory and work group barriers.

It maps to `shfl.sync.down.b32` in `PTX` for `CUDA`, see https://github.com/intel/llvm/blob/sycl/sycl/doc/cuda/opencl-subgroup-vs-cuda-crosslane-op.md.

So I believe the answer is "yes".

File: core/src/SYCL/Kokkos_SYCL_Team.hpp (hunk 10 above)
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Team.hpp:884]: `grange1` is the vector_size as requested and might be less than the subgroup size. Thus, `sg.get_local_id()` might be larger or equal than `grange1`.
Similarly, item.get_local_id(1) might be larger than `sg.get_local_id()` if there are multiple subgroups.

File: core/unit_test/incremental/Test12a_ThreadScratch.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/unit_test/incremental/Test12a_ThreadScratch.hpp:78]: See https://github.com/kokkos/kokkos/issues/4204.
